osu.Game/Skinning/LegacyDatabasedSkinResourceStore.cs

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using osu.Framework.Development;
+using osu.Framework.Extensions;
+using osu.Framework.IO.Stores;
+using osu.Game.Extensions;
+using Realms;
+
+namespace osu.Game.Skinning
+{
+    public class LegacyDatabasedSkinResourceStore : ResourceStore<byte[]>
+    {
+        private readonly Dictionary<string, string> fileToStoragePathMapping = new Dictionary<string, string>();
+
+        private readonly IDisposable subscription;
+
+        public LegacyDatabasedSkinResourceStore(SkinInfo source, IResourceStore<byte[]> underlyingStore)
+            : base(underlyingStore)
+        {
+            // Subscribing to non-managed instances doesn't work.
+            // In this usage, the skin may be non-managed in tests.
+            if (source.IsManaged)
+            {
+                // Subscriptions can only work on the main thread.
+                Debug.Assert(ThreadSafety.IsUpdateThread);
```

Comment:
Unfortunately this doesn't play well with tests, which are calling `SkinInfo.CreateInstance` in arbitrary thread contexts.

Almost tempted to leave out the subscription on this class until we need it. Right now it would never actually be required (as we don't change files in skins). At the point we are changing files, a refresh is likely going to be happening elsewhere at a higher level.

Interested in other opinions.

Reply:
If it isn't needed I'd be in favour of removing, especially because the finaliser added here (which is potentially disposing said subscription) is giving me the heebie-jeebies.

Reply:
I've removed this for now. Was mostly added as an example of how we'd handle this going forward anyways.

Note that the finaliser logic is generally how we do things when there are native components that require disposal (as a safety measure) so that shouldn't have set off any alarm bells, but that's an aside at this point.

Reply:
To be clear, it's more that by this point _every_ usage of finalisers is setting alarm bells in my head due to the lack of almost any guarantees they provide (to the point wherein I don't know if it's worth writing them even for safety), rather than this particular one being particularly alarming in one way or another.